kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Metadata.java

```diff
@@ -0,0 +1,156 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.types.ArrayType;
+import io.delta.kernel.types.LongType;
+import io.delta.kernel.types.MapType;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.types.StructType;
+
+import io.delta.kernel.internal.types.TableSchemaSerDe;
+
+public class Metadata implements Action
+{
+    public static Metadata fromRow(Row row, TableClient tableClient)
+    {
+        if (row == null) {
+            return null;
+        }
+
+        final String schemaJson = row.getString(4);
+        StructType schema = TableSchemaSerDe.fromJson(tableClient.getJsonHandler(), schemaJson);
+
+        return new Metadata(
+            row.getString(0),
+            Optional.ofNullable(row.isNullAt(1) ? null : row.getString(1)),
+            Optional.ofNullable(row.isNullAt(2) ? null : row.getString(2)),
+            Format.fromRow(row.getStruct(3)),
+            schemaJson,
+            schema,
+            row.getArray(5),
+            Optional.ofNullable(row.isNullAt(6) ? null : row.getLong(6)),
+            row.getMap(7)
+        );
+    }
+
+    public static final StructType READ_SCHEMA = new StructType()
+        .add("id", StringType.INSTANCE, false /* nullable */)
+        .add("name", StringType.INSTANCE, true /* nullable */)
+        .add("description", StringType.INSTANCE, true /* nullable */)
+        .add("format", Format.READ_SCHEMA, false /* nullable */)
+        .add("schemaString", StringType.INSTANCE, false /* nullable */)
+        .add("partitionColumns",
+            new ArrayType(StringType.INSTANCE, false /* contains null */),
+            false /* nullable */)
+        .add("createdTime", LongType.INSTANCE, true /* contains null */)
+        .add("configuration",
+            new MapType(StringType.INSTANCE, StringType.INSTANCE, false),
+            false /* contains null */);
+
+    private final String id;
+    private final Optional<String> name;
+    private final Optional<String> description;
+    private final Format format;
+    private final String schemaString;
+    private final StructType schema;
+    private final List<String> partitionColumns;
+    private final Optional<Long> createdTime;
+    private final Map<String, String> configuration;
+
+    public Metadata(
+        String id,
+        Optional<String> name,
+        Optional<String> description,
+        Format format,
+        String schemaString,
+        StructType schema,
+        List<String> partitionColumns,
+        Optional<Long> createdTime,
+        Map<String, String> configuration)
+    {
+        this.id = id;
+        this.name = name;
+        this.description = description;
+        this.format = format;
+        this.schemaString = schemaString;
+        this.schema = schema;
+        this.partitionColumns =
```

Comment:
for non-nullable arguments should we just `requireNonNull`? (same for other action constructors)